File: Core/src/main/java/com/hqq/core/net/ok/OkParamsImpl.java
/*
 * Create on 2016-10-26 下午2:47
 * FileName: OKParamsImpl.java
 * Author: Du Jianchi
 * Contact: du-1@qq.com or http://dujc.cn
 */

package com.hqq.core.net.ok;


import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.FileNameMap;
import java.net.URLConnection;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Headers;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

/**
 * @Author : huangqiqiang
 * @Package : com.qq.readbook.net.ok
 * @FileName :   OKParamsImpl
 * @Date : 2020/12/9 0009  下午 5:15
 * @Email :  qiqiang213@gmail.com
 * @Descrive :Created by du on 16/9/4.
 */
public class OkParamsImpl implements HttpCompat.ParamsCompat {

    private final List<String> keys = new ArrayList<>();
    private final List<Object> values = new ArrayList<>();

    @Override
    public HttpCompat.ParamsCompat put(String key, Object value) {
        keys.add(key);
        values.add(value);
        return this;
    }

    @Override
    public String paramGet() {
        return lineUp();
    }

    /**
     * @return
     */
    @Override
    public RequestBody paramForm() {
        return RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), lineUp());
    }

    /**
     * 表单方式提交数据
     * 正常是用于数据上传
     * 未测试
     *
     * @return
     */
    @Override
    public RequestBody paramMulti() {
        MultipartBody.Builder builder = new MultipartBody.Builder();
        builder.setType(MultipartBody.FORM);
        StringBuilder stringBuilder = new StringBuilder();
        final int length = keys.size();
        for (int index = 0; index < length; index++) {
            Object value = values.get(index);
            if (value instanceof File) {
                File f = (File) value;
                String filename = f.getName();
                RequestBody fileBody = RequestBody.create(MediaType.parse(guessMimeType(filename)), f);
                builder.addFormDataPart(keys.get(index), filename, fileBody);
                builder.addPart(fileBody);
            } else {
                builder.addPart(Headers.of("Content-Disposition", "form-data; name=\"" + keys.get(index) + "\""),
                        RequestBody.create(null, value + ""));
                stringBuilder.append(keys.get(index)).append('=').append(encode(value + "")).append('&');
            }
        }
        int len = stringBuilder.length();
        if (len > 0) {
            String strings = stringBuilder.substring(0, len - 1);
            RequestBody params = RequestBody.create(MultipartBody.FORM
                    , strings);
            builder.addPart(params);
        }
        return builder.build();
    }

    @Override
    public String toString() {
        return lineUp();
    }

    /**
     * 组装成一行数据
     *
     * @return
     */
    private String lineUp() {
        StringBuilder builder = new StringBuilder();
        final int length = keys.size();
        for (int index = 0; index < length - 1; index++) {
            builder.append(keys.get(index)).append('=').append(encode(values.get(index) + "")).append('&');
        }
        if (length - 1 >= 0) {
            builder.append(keys.get(length - 1)).append('=').append(encode(values.get(length - 1) + ""));
        }
        return builder.toString();
    }

    private static String encode(String string) {
        try {
            return URLEncoder.encode(string, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return string;
    }

    private static String guessMimeType(String path) {
        FileNameMap fileNameMap = URLConnection.getFileNameMap();
        String contentTypeFor = null;
        try {
            contentTypeFor = fileNameMap.getContentTypeFor(URLEncoder.encode(path, "UTF-8"));
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        if (contentTypeFor == null) {
            contentTypeFor = "application/octet-stream";
        }
        return contentTypeFor;
    }
}

File: Core/src/main/java/com/hqq/core/net/ok/OkParamsImpl.kt
/*
 * Create on 2016-10-26 下午2:47
 * FileName: OKParamsImpl.java
 * Author: Du Jianchi
 * Contact: du-1@qq.com or http://dujc.cn
 */
package com.hqq.core.net.ok

import com.hqq.core.net.ok.HttpCompat.ParamsCompat
import okhttp3.Headers
import okhttp3.Headers.Companion.headersOf
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File
import java.io.UnsupportedEncodingException
import java.net.URLConnection
import java.net.URLEncoder
import java.util.*
import kotlin.collections.HashMap

/**
 * @Author : huangqiqiang
 * @Package : com.qq.readbook.net.ok
 * @FileName :   OKParamsImpl
 * @Date : 2020/12/9 0009  下午 5:15
 * @Email :  qiqiang213@gmail.com
 * @Descrive :Created by du on 16/9/4.
 */
class OkParamsImpl : ParamsCompat {
    private val keys: MutableList<String?> = ArrayList()
    private val values: MutableList<Any?> = ArrayList()
    override var decode: String? = ""
    override var encode: String? = ""
    override var headers: MutableMap<String, String> =HashMap()


    override fun put(key: String?, value: Any?): ParamsCompat? {
        keys.add(key)
        values.add(value)
        return this
    }

    override fun paramGet(): String? {
        return lineUp()
    }

    override fun  paramForm(): RequestBody {
        return RequestBody.create("application/x-www-form-urlencoded".toMediaTypeOrNull(), lineUp())

    }

    /**
     * 表单方式提交数据
     * 正常是用于数据上传
     * 未测试
     *
     * @return
     */
    override fun paramMulti(): RequestBody {
        val builder = MultipartBody.Builder()
        builder.setType(MultipartBody.FORM)
        val stringBuilder = StringBuilder()
        val length = keys.size
        for (index in 0 until length) {
            val value = values[index]
            if (value is File) {
                val f = value
                val filename = f.name
                val fileBody = RequestBody.create(guessMimeType(filename).toMediaTypeOrNull(), f)
                keys[index]?.let { builder.addFormDataPart(it, filename, fileBody) }
                builder.addPart(fileBody)
            } else {
                builder.addPart(headersOf("Content-Disposition", "form-data; name=\"" + keys[index] + "\""),
                    RequestBody.create(null, value.toString() + ""))
                stringBuilder.append(keys[index]).append('=').append(encodeString(value.toString() + "")).append('&')
            }
        }
        val len = stringBuilder.length
        if (len > 0) {
            val strings = stringBuilder.substring(0, len - 1)
            val params = RequestBody.create(MultipartBody.FORM, strings)
            builder.addPart(params)
        }
        return builder.build()
    }

    override fun toString(): String {
        return lineUp()
    }

    /**
     * 组装成一行数据
     *
     * @return
     */
    private fun lineUp(): String {
        val builder = StringBuilder()
        val length = keys.size
        for (index in 0 until length - 1) {
            builder.append(keys[index]).append('=').append(encodeString(values[index].toString() + "")).append('&')
        }
        if (length - 1 >= 0) {
            builder.append(keys[length - 1]).append('=').append(encodeString(values[length - 1].toString() + ""))
        }
        return builder.toString()
    }

    private fun encodeString(string: String): String {
        if (encode == null || encode!!.isEmpty()) {
            return string
        }
        try {
            return URLEncoder.encode(string, encode)
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }
        return string
    }

    companion object {
        private fun guessMimeType(path: String): String {
            val fileNameMap = URLConnection.getFileNameMap()
            var contentTypeFor: String? = null
            try {
                contentTypeFor = fileNameMap.getContentTypeFor(URLEncoder.encode(path, "UTF-8"))
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
            }
            if (contentTypeFor == null) {
                contentTypeFor = "application/octet-stream"
            }
            return contentTypeFor
        }
    }
}